File: im-desktop/src/services/multimodalService.ts
/**
 * 多模态内容理解引擎 - TypeScript API 服务
 */

import {
  MultimodalConfig,
  AnalysisRequest,
  AnalysisResult,
  ConfigStats,
  ResultStats,
  AnalysisStatusResponse,
  HistoryItem,
  AnalysisQueryParams,
  SentimentDistribution,
  IntentDistribution,
  SceneDistribution,
  ApiResponse,
} from '../types/multimodal';

const API_BASE = '/api/multimodal';

/**
 * 多模态内容理解 API 服务
 */
class MultimodalService {
  private baseUrl: string;

  constructor(baseUrl: string = API_BASE) {
    this.baseUrl = baseUrl;
  }

  // ==================== HTTP 工具方法 ====================

  private async request<T>(
    endpoint: string,
    method: string = 'GET',
    body?: any
  ): Promise<ApiResponse<T>> {
    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
      },
    };

    if (body) {
      options.body = JSON.stringify(body);
    }

    const response = await fetch(`${this.baseUrl}${endpoint}`, options);
    
    if (!response.ok) {
      const errorData = await response.json().catch(() => ({ message: '请求失败' }));
      throw new Error(errorData.message || `HTTP ${response.status}`);
    }

    return response.json();
  }

  // ==================== 配置管理 API ====================

  /**
   * 创建新的多模态配置
   */
  async createConfig(config: MultimodalConfig): Promise<MultimodalConfig> {
    const response = await this.request<MultimodalConfig>('/configs', 'POST', config);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 更新配置
   */
  async updateConfig(id: number, config: Partial<MultimodalConfig>): Promise<MultimodalConfig> {
    const response = await this.request<MultimodalConfig>(`/configs/${id}`, 'PUT', config);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 根据 ID 获取配置
   */
  async getConfigById(id: number): Promise<MultimodalConfig> {
    const response = await this.request<MultimodalConfig>(`/configs/${id}`);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 根据名称获取配置
   */
  async getConfigByName(name: string): Promise<MultimodalConfig> {
    const response = await this.request<MultimodalConfig>(`/configs/name/${name}`);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 获取所有启用的配置
   */
  async getAllEnabledConfigs(): Promise<MultimodalConfig[]> {
    const response = await this.request<MultimodalConfig[]>('/configs/enabled');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data || [];
  }

  /**
   * 获取默认配置
   */
  async getDefaultConfig(): Promise<MultimodalConfig> {
    const response = await this.request<MultimodalConfig>('/configs/default');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 禁用配置
   */
  async disableConfig(id: number): Promise<void> {
    await this.request(`/configs/${id}/disable`, 'POST');
  }

  /**
   * 启用配置
   */
  async enableConfig(id: number): Promise<void> {
    await this.request(`/configs/${id}/enable`, 'POST');
  }

  /**
   * 删除配置
   */
  async deleteConfig(id: number): Promise<void> {
    await this.request(`/configs/${id}`, 'DELETE');
  }

  /**
   * 获取配置统计信息
   */
  async getConfigStats(): Promise<ConfigStats> {
    const response = await this.request<ConfigStats>('/configs/stats');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  // ==================== 分析结果 API ====================

  /**
   * 提交分析请求
   */
  async analyze(request: AnalysisRequest): Promise<{ requestId: string; status: string }> {
    const response = await this.request<{ requestId: string; status: string }>('/analyze', 'POST', request);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 获取分析结果状态
   */
  async getAnalysisStatus(requestId: string): Promise<AnalysisStatusResponse> {
    const response = await this.request<AnalysisStatusResponse>(`/results/${requestId}/status`);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 获取分析结果详情
   */
  async getAnalysisResult(requestId: string): Promise<AnalysisResult | { status: string }> {
    const response = await this.request<AnalysisResult>(`/results/${requestId}`);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 根据消息 ID 获取分析结果
   */
  async getAnalysisResultByMessageId(messageId: number): Promise<AnalysisResult | null> {
    const response = await this.request<AnalysisResult | null>(`/results/by-message/${messageId}`);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data || null;
  }

  /**
   * 获取用户分析历史
   */
  async getUserAnalysisHistory(userId: number, limit: number = 20): Promise<HistoryItem[]> {
    const response = await this.request<HistoryItem[]>(`/users/${userId}/history?limit=${limit}`);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data || [];
  }

  /**
   * 获取会话分析历史
   */
  async getSessionAnalysisHistory(sessionId: string, limit: number = 20): Promise<HistoryItem[]> {
    const response = await this.request<HistoryItem[]>(`/sessions/${sessionId}/history?limit=${limit}`);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data || [];
  }

  /**
   * 获取分析结果统计
   */
  async getResultStats(): Promise<ResultStats> {
    const response = await this.request<ResultStats>('/results/stats');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 批量查询分析结果
   */
  async searchResults(params: AnalysisQueryParams): Promise<{
    page: number;
    size: number;
    total: number;
    results: HistoryItem[];
  }> {
    const queryParams = new URLSearchParams();
    
    if (params.userId) queryParams.append('userId', params.userId.toString());
    if (params.contentType) queryParams.append('contentType', params.contentType);
    if (params.status) queryParams.append('status', params.status);
    if (params.startDate) queryParams.append('startDate', params.startDate);
    if (params.endDate) queryParams.append('endDate', params.endDate);
    queryParams.append('page', params.page?.toString() || '0');
    queryParams.append('size', params.size?.toString() || '20');
    queryParams.append('sortBy', params.sortBy || 'createdAt');
    queryParams.append('sortDir', params.sortDir || 'DESC');

    const response = await this.request(`/results?${queryParams.toString()}`);
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  // ==================== 系统管理 API ====================

  /**
   * 清理过期缓存
   */
  async cleanupExpiredCache(): Promise<{ cleanedCount: number; timestamp: string }> {
    const response = await this.request('/admin/cleanup-cache', 'POST');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  /**
   * 重试失败请求
   */
  async retryFailedRequests(): Promise<{ retriedCount: number; requests: Array<{ requestId: string; retryCount: number }> }> {
    const response = await this.request('/admin/retry-failed', 'POST');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  // ==================== 统计分析 API ====================

  /**
   * 获取情感分布
   */
  async getSentimentDistribution(): Promise<SentimentDistribution[]> {
    const response = await this.request<SentimentDistribution[]>('/stats/sentiment-distribution');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data || [];
  }

  /**
   * 获取意图分布
   */
  async getIntentDistribution(): Promise<IntentDistribution[]> {
    const response = await this.request<IntentDistribution[]>('/stats/intent-distribution');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data || [];
  }

  /**
   * 获取场景分布
   */
  async getSceneDistribution(): Promise<SceneDistribution[]> {
    const response = await this.request<SceneDistribution[]>('/stats/scene-distribution');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data || [];
  }

  /**
   * 系统健康检查
   */
  async healthCheck(): Promise<{
    status: string;
    timestamp: string;
    configStats: ConfigStats;
    resultStats: ResultStats;
    message: string;
  }> {
    const response = await this.request('/health');
    if (!response.success) {
      throw new Error(response.message);
    }
    return response.data!;
  }

  // ==================== 辅助方法 ====================

  /**
   * 轮询分析结果直到完成
   */
  async pollUntilComplete(
    requestId: string,
    intervalMs: number = 1000,
    timeoutMs: number = 60000
  ): Promise<AnalysisResult> {
    const startTime = Date.now();

    while (Date.now() - startTime < timeoutMs) {
      const status = await this.getAnalysisStatus(requestId);

      if (status.status === 'completed') {
        const result = await this.getAnalysisResult(requestId);
        if ('analysisStatus' in result) {
          return result;
        }
        throw new Error('分析结果格式异常');
      }

      if (status.status === 'failed') {
        throw new Error(`分析失败: ${status.errorMessage}`);
      }

      // 等待下一次轮询
      await new Promise(resolve => setTimeout(resolve, intervalMs));
    }

    throw new Error('分析超时');
  }

  /**
   * 分析文本内容
   */
  async analyzeText(
    text: string,
    options?: {
      sessionId?: string;
      userId?: number;
      messageId?: number;
      priority?: number;
    }
  ): Promise<string> {
    const response = await this.analyze({
      contentType: 'text',
      textContent: text,
      ...options,
    });
    return response.requestId;
  }

  /**
   * 分析图像内容
   */
  async analyzeImage(
    imageUrl: string,
    options?: {
      sessionId?: string;
      userId?: number;
      messageId?: number;
      priority?: number;
    }
  ): Promise<string> {
    const response = await this.analyze({
      contentType: 'image',
      imageUrl,
      ...options,
    });
    return response.requestId;
  }

  /**
   * 分析音频内容
   */
  async analyzeAudio(
    audioUrl: string,
    options?: {
      sessionId?: string;
      userId?: number;
      messageId?: number;
      priority?: number;
    }
  ): Promise<string> {
    const response = await this.analyze({
      contentType: 'audio',
      audioUrl,
      ...options,
    });
    return response.requestId;
  }

  /**
   * 分析视频内容
   */
  async analyzeVideo(
    videoUrl: string,
    options?: {
      sessionId?: string;
      userId?: number;
      messageId?: number;
      priority?: number;
    }
  ): Promise<string> {
    const response = await this.analyze({
      contentType: 'video',
      videoUrl,
      ...options,
    });
    return response.requestId;
  }

  /**
   * 分析多模态内容
   */
  async analyzeMultimodal(
    content: {
      text?: string;
      imageUrl?: string;
      audioUrl?: string;
      videoUrl?: string;
    },
    options?: {
      sessionId?: string;
      userId?: number;
      messageId?: number;
      priority?: number;
      businessContext?: string;
    }
  ): Promise<string> {
    const response = await this.analyze({
      contentType: 'mixed',
      ...content,
      ...options,
    });
    return response.requestId;
  }
}

// 导出单例
export const multimodalService = new MultimodalService();
export default multimodalService;

File: im-mobile/lib/services/multimodal_service.dart
/**
 * 多模态内容理解引擎 - Dart Flutter 服务
 */

import 'dart:convert';
import 'package:http/http.dart' as http;

// ==================== 数据模型 ====================

/// 多模态配置
class MultimodalConfig {
  final int? id;
  final String name;
  final String? description;
  final bool enabled;
  final int? version;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  
  // 文本理解配置
  final bool textEnabled;
  final String? textModel;
  final int? textMaxLength;
  final String? textLanguages;
  
  // 图像理解配置
  final bool imageEnabled;
  final String? imageModel;
  final int? imageMaxSize;
  final String? imageSupportedFormats;
  
  // 语音理解配置
  final bool audioEnabled;
  final String? audioModel;
  final int? audioMaxDuration;
  final String? audioSupportedFormats;
  
  // 视频理解配置
  final bool videoEnabled;
  final String? videoModel;
  final int? videoMaxDuration;
  final int? videoMaxSize;
  
  // 多模态融合配置
  final bool multimodalFusionEnabled;
  final String? fusionMethod;
  final String? crossModalWeighting;
  
  // 缓存配置
  final bool cacheEnabled;
  final int? cacheTtlHours;
  final int? cacheMaxSize;
  
  // 性能配置
  final int? concurrentWorkers;
  final int? timeoutMs;
  final int? batchSize;
  
  // 质量配置
  final double? confidenceThreshold;
  final bool? fallbackEnabled;
  final String? fallbackModel;
  
  // 监控配置
  final bool metricsEnabled;
  final int? metricsIntervalMinutes;
  final double? alertThresholdErrorRate;
  final int? alertThresholdLatencyMs;
  
  // 隐私配置
  final bool privacyEnabled;
  final bool? anonymizationEnabled;
  final int? dataRetentionDays;
  
  // 自定义配置
  final Map<String, dynamic>? customConfig;

  MultimodalConfig({
    this.id,
    required this.name,
    this.description,
    required this.enabled,
    this.version,
    this.createdAt,
    this.updatedAt,
    required this.textEnabled,
    this.textModel,
    this.textMaxLength,
    this.textLanguages,
    required this.imageEnabled,
    this.imageModel,
    this.imageMaxSize,
    this.imageSupportedFormats,
    required this.audioEnabled,
    this.audioModel,
    this.audioMaxDuration,
    this.audioSupportedFormats,
    required this.videoEnabled,
    this.videoModel,
    this.videoMaxDuration,
    this.videoMaxSize,
    required this.multimodalFusionEnabled,
    this.fusionMethod,
    this.crossModalWeighting,
    required this.cacheEnabled,
    this.cacheTtlHours,
    this.cacheMaxSize,
    this.concurrentWorkers,
    this.timeoutMs,
    this.batchSize,
    this.confidenceThreshold,
    this.fallbackEnabled,
    this.fallbackModel,
    required this.metricsEnabled,
    this.metricsIntervalMinutes,
    this.alertThresholdErrorRate,
    this.alertThresholdLatencyMs,
    required this.privacyEnabled,
    this.anonymizationEnabled,
    this.dataRetentionDays,
    this.customConfig,
  });

  factory MultimodalConfig.fromJson(Map<String, dynamic> json) {
    return MultimodalConfig(
      id: json['id'],
      name: json['name'] ?? '',
      description: json['description'],
      enabled: json['enabled'] ?? false,
      version: json['version'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      textEnabled: json['textEnabled'] ?? false,
      textModel: json['textModel'],
      textMaxLength: json['textMaxLength'],
      textLanguages: json['textLanguages'],
      imageEnabled: json['imageEnabled'] ?? false,
      imageModel: json['imageModel'],
      imageMaxSize: json['imageMaxSize'],
      imageSupportedFormats: json['imageSupportedFormats'],
      audioEnabled: json['audioEnabled'] ?? false,
      audioModel: json['audioModel'],
      audioMaxDuration: json['audioMaxDuration'],
      audioSupportedFormats: json['audioSupportedFormats'],
      videoEnabled: json['videoEnabled'] ?? false,
      videoModel: json['videoModel'],
      videoMaxDuration: json['videoMaxDuration'],
      videoMaxSize: json['videoMaxSize'],
      multimodalFusionEnabled: json['multimodalFusionEnabled'] ?? false,
      fusionMethod: json['fusionMethod'],
      crossModalWeighting: json['crossModalWeighting'],
      cacheEnabled: json['cacheEnabled'] ?? false,
      cacheTtlHours: json['cacheTtlHours'],
      cacheMaxSize: json['cacheMaxSize'],
      concurrentWorkers: json['concurrentWorkers'],
      timeoutMs: json['timeoutMs'],
      batchSize: json['batchSize'],
      confidenceThreshold: json['confidenceThreshold']?.toDouble(),
      fallbackEnabled: json['fallbackEnabled'],
      fallbackModel: json['fallbackModel'],
      metricsEnabled: json['metricsEnabled'] ?? false,
      metricsIntervalMinutes: json['metricsIntervalMinutes'],
      alertThresholdErrorRate: json['alertThresholdErrorRate']?.toDouble(),
      alertThresholdLatencyMs: json['alertThresholdLatencyMs'],
      privacyEnabled: json['privacyEnabled'] ?? false,
      anonymizationEnabled: json['anonymizationEnabled'],
      dataRetentionDays: json['dataRetentionDays'],
      customConfig: json['customConfig'] != null 
          ? Map<String, dynamic>.from(json['customConfig']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'enabled': enabled,
      'version': version,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'textEnabled': textEnabled,
      'textModel': textModel,
      'textMaxLength': textMaxLength,
      'textLanguages': textLanguages,
      'imageEnabled': imageEnabled,
      'imageModel': imageModel,
      'imageMaxSize': imageMaxSize,
      'imageSupportedFormats': imageSupportedFormats,
      'audioEnabled': audioEnabled,
      'audioModel': audioModel,
      'audioMaxDuration': audioMaxDuration,
      'audioSupportedFormats': audioSupportedFormats,
      'videoEnabled': videoEnabled,
      'videoModel': videoModel,
      'videoMaxDuration': videoMaxDuration,
      'videoMaxSize': videoMaxSize,
      'multimodalFusionEnabled': multimodalFusionEnabled,
      'fusionMethod': fusionMethod,
      'crossModalWeighting': crossModalWeighting,
      'cacheEnabled': cacheEnabled,
      'cacheTtlHours': cacheTtlHours,
      'cacheMaxSize': cacheMaxSize,
      'concurrentWorkers': concurrentWorkers,
      'timeoutMs': timeoutMs,
      'batchSize': batchSize,
      'confidenceThreshold': confidenceThreshold,
      'fallbackEnabled': fallbackEnabled,
      'fallbackModel': fallbackModel,
      'metricsEnabled': metricsEnabled,
      'metricsIntervalMinutes': metricsIntervalMinutes,
      'alertThresholdErrorRate': alertThresholdErrorRate,
      'alertThresholdLatencyMs': alertThresholdLatencyMs,
      'privacyEnabled': privacyEnabled,
      'anonymizationEnabled': anonymizationEnabled,
      'dataRetentionDays': dataRetentionDays,
      'customConfig': customConfig,
    };
  }
}

/// 分析请求
class AnalysisRequest {
  final String? sessionId;
  final int? userId;
  final int? messageId;
  final String contentType;
  final String? businessContext;
  final int? priority;
  final String? textContent;
  final String? imageUrl;
  final String? audioUrl;
  final String? videoUrl;

  AnalysisRequest({
    this.sessionId,
    this.userId,
    this.messageId,
    required this.contentType,
    this.businessContext,
    this.priority,
    this.textContent,
    this.imageUrl,
    this.audioUrl,
    this.videoUrl,
  });

  Map<String, dynamic> toJson() {
    return {
      'sessionId': sessionId,
      'userId': userId,
      'messageId': messageId,
      'contentType': contentType,
      'businessContext': businessContext,
      'priority': priority,
      'textContent': textContent,
      'imageUrl': imageUrl,
      'audioUrl': audioUrl,
      'videoUrl': videoUrl,
    };
  }
}

/// 分析结果
class AnalysisResult {
  final String requestId;
  final String? sessionId;
  final int? userId;
  final int? messageId;
  final String contentType;
  final String? contentHash;
  final String analysisStatus;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? completedAt;
  final Map<String, dynamic>? textAnalysis;
  final Map<String, dynamic>? imageAnalysis;
  final Map<String, dynamic>? audioAnalysis;
  final Map<String, dynamic>? videoAnalysis;
  final Map<String, dynamic>? multimodalFusion;
  final double? confidenceScore;
  final String? qualityRating;
  final int? processingTimeMs;
  final String? modelUsed;
  final double? costUnits;
  final String? errorMessage;
  final String? errorCode;
  final int? retryCount;

  AnalysisResult({
    required this.requestId,
    this.sessionId,
    this.userId,
    this.messageId,
    required this.contentType,
    this.contentHash,
    required this.analysisStatus,
    required this.createdAt,
    required this.updatedAt,
    this.completedAt,
    this.textAnalysis,
    this.imageAnalysis,
    this.audioAnalysis,
    this.videoAnalysis,
    this.multimodalFusion,
    this.confidenceScore,
    this.qualityRating,
    this.processingTimeMs,
    this.modelUsed,
    this.costUnits,
    this.errorMessage,
    this.errorCode,
    this.retryCount,
  });

  factory AnalysisResult.fromJson(Map<String, dynamic> json) {
    return AnalysisResult(
      requestId: json['requestId'] ?? '',
      sessionId: json['sessionId'],
      userId: json['userId'],
      messageId: json['messageId'],
      contentType: json['contentType'] ?? '',
      contentHash: json['contentHash'],
      analysisStatus: json['analysisStatus'] ?? '',
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null 
          ? DateTime.parse(json['updatedAt']) 
          : DateTime.now(),
      completedAt: json['completedAt'] != null 
          ? DateTime.parse(json['completedAt']) 
          : null,
      textAnalysis: json['textAnalysis'] != null 
          ? Map<String, dynamic>.from(json['textAnalysis']) 
          : null,
      imageAnalysis: json['imageAnalysis'] != null 
          ? Map<String, dynamic>.from(json['imageAnalysis']) 
          : null,
      audioAnalysis: json['audioAnalysis'] != null 
          ? Map<String, dynamic>.from(json['audioAnalysis']) 
          : null,
      videoAnalysis: json['videoAnalysis'] != null 
          ? Map<String, dynamic>.from(json['videoAnalysis']) 
          : null,
      multimodalFusion: json['multimodalFusion'] != null 
          ? Map<String, dynamic>.from(json['multimodalFusion']) 
          : null,
      confidenceScore: json['confidenceScore']?.toDouble(),
      qualityRating: json['qualityRating'],
      processingTimeMs: json['processingTimeMs'],
      modelUsed: json['modelUsed'],
      costUnits: json['costUnits']?.toDouble(),
      errorMessage: json['errorMessage'],
      errorCode: json['errorCode'],
      retryCount: json['retryCount'],
    );
  }

  bool get isCompleted => analysisStatus == 'completed';
  bool get isFailed => analysisStatus == 'failed';
  bool get isProcessing => analysisStatus == 'processing';
  bool get isPending => analysisStatus == 'pending';
}

/// 配置统计
class ConfigStats {
  final int totalConfigs;
  final int enabledConfigs;
  final int textAnalysisConfigs;
  final int imageAnalysisConfigs;
  final int audioAnalysisConfigs;
  final int videoAnalysisConfigs;
  final int fallbackEnabledConfigs;

  ConfigStats({
    required this.totalConfigs,
    required this.enabledConfigs,
    required this.textAnalysisConfigs,
    required this.imageAnalysisConfigs,
    required this.audioAnalysisConfigs,
    required this.videoAnalysisConfigs,
    required this.fallbackEnabledConfigs,
  });

  factory ConfigStats.fromJson(Map<String, dynamic> json) {
    return ConfigStats(
      totalConfigs: json['totalConfigs'] ?? 0,
      enabledConfigs: json['enabledConfigs'] ?? 0,
      textAnalysisConfigs: json['textAnalysisConfigs'] ?? 0,
      imageAnalysisConfigs: json['imageAnalysisConfigs'] ?? 0,
      audioAnalysisConfigs: json['audioAnalysisConfigs'] ?? 0,
      videoAnalysisConfigs: json['videoAnalysisConfigs'] ?? 0,
      fallbackEnabledConfigs: json['fallbackEnabledConfigs'] ?? 0,
    );
  }
}

/// 结果统计
class ResultStats {
  final int totalResults;
  final int completedResults;
  final int failedResults;
  final int processingResults;
  final int pendingResults;
  final double? averageProcessingTime;
  final double? averageConfidenceScore;
  final double? averageSentimentScore;
  final double? totalCost;

  ResultStats({
    required this.totalResults,
    required this.completedResults,
    required this.failedResults,
    required this.processingResults,
    required this.pendingResults,
    this.averageProcessingTime,
    this.averageConfidenceScore,
    this.averageSentimentScore,
    this.totalCost,
  });

  factory ResultStats.fromJson(Map<String, dynamic> json) {
    return ResultStats(
      totalResults: json['totalResults'] ?? 0,
      completedResults: json['completedResults'] ?? 0,
      failedResults: json['failedResults'] ?? 0,
      processingResults: json['processingResults'] ?? 0,
      pendingResults: json['pendingResults'] ?? 0,
      averageProcessingTime: json['averageProcessingTime']?.toDouble(),
      averageConfidenceScore: json['averageConfidenceScore']?.toDouble(),
      averageSentimentScore: json['averageSentimentScore']?.toDouble(),
      totalCost: json['totalCost']?.toDouble(),
    );
  }
}

// ==================== API 服务 ====================

class MultimodalApiService {
  final String baseUrl;
  final http.Client _client;

  MultimodalApiService({
    this.baseUrl = '/api/multimodal',
    http.Client? client,
  }) : _client = client ?? http.Client();

  Future<Map<String, dynamic>> _request(
    String endpoint, {
    String method = 'GET',
    Map<String, dynamic>? body,
  }) async {
    final url = Uri.parse('$baseUrl$endpoint');
    
    final response = await _client.request(
      method,
      url,
      headers: {'Content-Type': 'application/json'},
      body: body != null ? jsonEncode(body) : null,
    );

    if (response.statusCode != 200) {
      final errorData = jsonDecode(response.body);
      throw Exception(errorData['message'] ?? '请求失败');
    }

    final responseData = jsonDecode(response.body);
    if (!responseData['success']) {
      throw Exception(responseData['message'] ?? '请求失败');
    }

    return responseData;
  }

  // ==================== 配置管理 ====================

  Future<MultimodalConfig> createConfig(MultimodalConfig config) async {
    final response = await _request('/configs', method: 'POST', body: config.toJson());
    return MultimodalConfig.fromJson(response['data']);
  }

  Future<MultimodalConfig> updateConfig(int id, Map<String, dynamic> updates) async {
    final response = await _request('/configs/$id', method: 'PUT', body: updates);
    return MultimodalConfig.fromJson(response['data']);
  }

  Future<MultimodalConfig> getConfigById(int id) async {
    final response = await _request('/configs/$id');
    return MultimodalConfig.fromJson(response['data']);
  }

  Future<MultimodalConfig> getConfigByName(String name) async {
    final response = await _request('/configs/name/$name');
    return MultimodalConfig.fromJson(response['data']);
  }

  Future<List<MultimodalConfig>> getAllEnabledConfigs() async {
    final response = await _request('/configs/enabled');
    final data = response['data'] as List;
    return data.map((item) => MultimodalConfig.fromJson(item)).toList();
  }

  Future<MultimodalConfig> getDefaultConfig() async {
    final response = await _request('/configs/default');
    return MultimodalConfig.fromJson(response['data']);
  }

  Future<void> disableConfig(int id) async {
    await _request('/configs/$id/disable', method: 'POST');
  }

  Future<void> enableConfig(int id) async {
    await _request('/configs/$id/enable', method: 'POST');
  }

  Future<void> deleteConfig(int id) async {
    await _request('/configs/$id', method: 'DELETE');
  }

  Future<ConfigStats> getConfigStats() async {
    final response = await _request('/configs/stats');
    return ConfigStats.fromJson(response['data']);
  }

  // ==================== 分析结果 ====================

  Future<Map<String, dynamic>> analyze(AnalysisRequest request) async {
    final response = await _request('/analyze', method: 'POST', body: request.toJson());
    return response['data'];
  }

  Future<Map<String, dynamic>> getAnalysisStatus(String requestId) async {
    final response = await _request('/results/$requestId/status');
    return response['data'];
  }

  Future<AnalysisResult> getAnalysisResult(String requestId) async {
    final response = await _request('/results/$requestId');
    return AnalysisResult.fromJson(response['data']);
  }

  Future<AnalysisResult?> getAnalysisResultByMessageId(int messageId) async {
    final response = await _request('/results/by-message/$messageId');
    final data = response['data'];
    if (data == null) return null;
    return AnalysisResult.fromJson(data);
  }

  Future<List<Map<String, dynamic>>> getUserAnalysisHistory(
    int userId, {
    int limit = 20,
  }) async {
    final response = await _request('/users/$userId/history?limit=$limit');
    final data = response['data'] as List;
    return data.cast<Map<String, dynamic>>();
  }

  Future<List<Map<String, dynamic>>> getSessionAnalysisHistory(
    String sessionId, {
    int limit = 20,
  }) async {
    final response = await _request('/sessions/$sessionId/history?limit=$limit');
    final data = response['data'] as List;
    return data.cast<Map<String, dynamic>>();
  }

  Future<ResultStats> getResultStats() async {
    final response = await _request('/results/stats');
    return ResultStats.fromJson(response['data']);
  }

  Future<Map<String, dynamic>> cleanupExpiredCache() async {
    final response = await _request('/admin/cleanup-cache', method: 'POST');
    return response['data'];
  }

  Future<Map<String, dynamic>> retryFailedRequests() async {
    final response = await _request('/admin/retry-failed', method: 'POST');
    return response['data'];
  }

  // ==================== 便捷方法 ====================

  Future<String> analyzeText(
    String text, {
    String? sessionId,
    int? userId,
    int? messageId,
    int? priority,
  }) async {
    final response = await analyze(AnalysisRequest(
      contentType: 'text',
      textContent: text,
      sessionId: sessionId,
      userId: userId,
      messageId: messageId,
      priority: priority,
    ));
    return response['requestId'];
  }

  Future<String> analyzeImage(
    String imageUrl, {
    String? sessionId,
    int? userId,
    int? messageId,
    int? priority,
  }) async {
    final response = await analyze(AnalysisRequest(
      contentType: 'image',
      imageUrl: imageUrl,
      sessionId: sessionId,
      userId: userId,
      messageId: messageId,
      priority: priority,
    ));
    return response['requestId'];
  }

  Future<String> analyzeAudio(
    String audioUrl, {
    String? sessionId,
    int? userId,
    int? messageId,
    int? priority,
  }) async {
    final response = await analyze(AnalysisRequest(
      contentType: 'audio',
      audioUrl: audioUrl,
      sessionId: sessionId,
      userId: userId,
      messageId: messageId,
      priority: priority,
    ));
    return response['requestId'];
  }

  Future<String> analyzeVideo(
    String videoUrl, {
    String? sessionId,
    int? userId,
    int? messageId,
    int? priority,
  }) async {
    final response = await analyze(AnalysisRequest(
      contentType: 'video',
      videoUrl: videoUrl,
      sessionId: sessionId,
      userId: userId,
      messageId: messageId,
      priority: priority,
    ));
    return response['requestId'];
  }

  void dispose() {
    _client.close();
  }
}